[ADD] stock_measuring_device[_zippcube[_reception_screen]]

carlosserratoro opened the pull request:
On-going development. Second 'clean' version, prepared to be refactored. And refactored a bit now.

Re-does the original module `stock_zipcube`, proposed in the PR https://github.com/OCA/stock-logistics-warehouse/pull/1096 so that it doesn't depend on `stock_cubiscan`.

The measuring has been made available from the reception screen, and this is implemented into the new module with name `stock_measuring_device_zippcube_reception_screen`, that maybe could go into the repository `wms`, since `reception_screen` is from there.

Commit:
[ADD] stock_measuring_device_zippcube[_reception_screen]

On-going development. First 'clean' version, not yet
preared to be refactored.

Re-does the original module stock_zipcube, proposed in the PR
https://github.com/OCA/stock-logistics-warehouse/pull/1096
so that it doesn't depend on stock_cubiscan. For the moment,
stock_measuring_device_zippcube depends on itself, but the idea
is to extract the common parts for other measuring devices,
such as Cubiscan, and put them in some other module.

The measuring has been made available from the reception screen,
and this is implemented into the new module with name
stock_measuring_device_zippcube_reception_screen, that maybe
could go into the repository wms, since reception_screen is
from there.

damdam-s commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -10,26 +10,39 @@
+        )
+        if not zippcube:
+            raise MissingError(
+                _(f"No such Zippcube with name {zippcube_device_name}.")
```
f-strings doesn't translate well. see https://stackoverflow.com/questions/49797658/how-to-use-gettext-with-python-3-6-f-strings

carlosserratoro replied:
Thanks, I didn't know. I changed all f-strings to use `format` instead.

damdam-s commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -10,26 +10,39 @@
+
+        keys_missing = set(self.expected_keys) - set(data)
+        if keys_missing:
+            error_msg = _(f"Wrong data format: {data}. Keys missing: {keys_missing}.")
```
same for translation

damdam-s commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+        )
+
+        if len(packaging) == 0:
+            error_msg = _(
```
same about f-string translation

damdam-s commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+            raise UserError(error_msg)
+
+        elif len(packaging) > 1:
+            warning_msg = _(
```
same about f-string translation

damdam-s commented on stock_measuring_device_zippcube/readme/ROADMAP.rst:
```diff
@@ -0,0 +1,6 @@
+* There are other measuring and weighing devices in the marked, and
```
```suggestion
* There are other measuring and weighing devices in the market, and
```

damdam-s commented on stock_measuring_device_zippcube/wizard/zippcube_wizard.py:
```diff
@@ -0,0 +1,182 @@
+        vals_list = []
+        packaging_types = self.env["product.packaging.type"].search([])
+        for seq, pack_type in enumerate(packaging_types):
+            pack = self.env["product.packaging"].search(
```
```suggestion
            pack = product_packaging.search(
```
and put `product_packaging = self.env["product.packaging"]` outside of the loop

damdam-s reviewed (changes requested):
- handle translation properly in the code
- missing translation (if needed)

simahawk commented on stock_measuring_device_zippcube/__manifest__.py:
```diff
@@ -0,0 +1,33 @@
+# Copyright 2021 Camptocamp SA
```
I'd rename to `stock_measure_device_*` and I'd add already `stock_measure_device` w/ just a boiler plate of mixin to use. You will move piece of codes to them later.

simahawk commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -10,26 +10,39 @@
-        if not cubiscan:
-            raise MissingError(_("No such cubiscan"))
+    @http.route(
+        "/stock/zippcube/<string:zippcube_device_name>/measurement",
```
why not `<model(zippcube.device):device>`? You'll have the object straight using its slug.

carlosserratoro replied:
I don't find any documentation on how this works. Does it search by the model's name?

simahawk commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -10,26 +10,39 @@
         # convert the float values passed as strings to floats
         data = self._convert_floats(data)
-        cubiscan._update_packaging_measures(data)
+        zippcube._update_packaging_measures(data)
```
call it `device` everywhere, it will be easier to make refactor things for generic usage.

simahawk commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -44,12 +57,12 @@ def _convert_floats(self, data):
         return data
 
     def _check_secret(self, secret):
-        if secret and secret == os.environ.get("ZIPCUBE_SECRET"):
+        if secret and secret == os.environ.get("ZIPPCUBE_SECRET"):
```
```py
DEVICE_SECRET_KEY = "ZIPPCUBE_SECRET"
[...]
def _device_get_secret(self):
    return os.environ.get(self.DEVICE_SECRET_KEY)
[...]
def _check_secret(self, secret):
    if secret and secret == self._device_get_secret():
[...]
```

simahawk commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+        A device is in use if it has been assigned to a product packaging."""
+        return bool(
+            self.env["product.packaging"].search(
+                [("scan_device_id", "=", self.id)], count=True, limit=1
```
`limit` is useless if you use `count` no?

carlosserratoro replied:
You are right... I change it to `search_count` then. Thanks. (My idea here was to check just for existence, but `LIMIT` doesn't stop scanning).

simahawk commented on stock_measuring_device_zippcube/models/product_packaging.py:
```diff
@@ -0,0 +1,28 @@
+class ProductPackaging(models.Model):
+    _inherit = "product.packaging"
+
+    scan_device_id = fields.Many2one(
```
this can't have a generic name if it's tied to a specific model.
Unless you plan to have a generic model `measurement.device`

simahawk commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+    _description = "Measuring Device Zippcube"
+    _order = "warehouse_id, name"
+
+    name = fields.Char("Name", required=True)
```
almost everything here can stay in a mixin and/or common model.
You can have a `device_type_id` and based on the details of the type you can provide all the configuration you need (name, description, wizard reference, etc)

simahawk commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+            [("scan_device_id", "=", self.id)], order="write_date DESC"
+        )
+
+        if len(packaging) == 0:
```
split this to `_validate_packaging`

simahawk commented on stock_measuring_device_zippcube/models/zippcube_device.py:
```diff
@@ -0,0 +1,109 @@
+        if not wizard_line:
+            _logger.warning("No wizard line found for this measure.")
+            packaging_to_update.write(
+                {
```
vals hare the same as write below, you can define them before

simahawk commented:
I fail to get why there's no base for the common module. I'm not saying everything should be generic but at least a draft of it. Otherwise, very likely, in 2 months from now you will be still here to refactor things.
But probably I miss some pieces of the puzzle :wink:

Commit:
[IMP] stock_measuring_device[_zippcube[_reception_screen]]

- Changed the f-strings. One reviewer commented that they
  are not translatable. I now use the old reliable format().
- A new module stock_measuring_device has been added with
  the common parts: the wizard is now common for any
  measuring device, or at least intends to be.
- Pending: Making the model measuring.device an abstract
  model to create new specific measuring devices, such as
  zippcube.

Commit:
[WIP] Don't merge. It doesn't work.

This is a work in progress. It's intention was to move to the
use of components so that measuring.device was implemented using
them. The previous commit to this one is the one that works,
but doesn't use components (but simple 2nd-type of inheritance).
